include/SingleCellProtocols.hpp

```diff
@@ -220,7 +221,49 @@ namespace alevin{
       BarcodeEnd end;
     };
 
+    // Custom geometry specification using regex for extraction
+    struct CustomGeo {
+      // store regex string for reads 1 and 2
+      static std::string reg[2];
+      // store positions of matches for bc and umi
+      static std::vector<int> b[2], u[2];
```

Comment:
Having the regex themselves as static makes sense, I think, because the share shared over all reads and read-only.  However, should we be concerned having a static vector of match positions? Doesn't this mean it's shared over all instances of the class (cc @gmarcais)?  Is it therefore not threadsafe to extract UMI & BC from multiple threads at the same time?  That would be bad.

Also, given that we're using a vector here that we expect, in general, to have a very small number of values, I would suggest we use a vector class with small vector optimization (with storage on the stack below a certain # of elements).  We already use the `chobo::small_vector` in this codebase, so we could use that here.  However, looking at it it now seems [that library](https://github.com/Chobolabs/chobo-shl) is no longer actively supported and the same developer has moved all the functionality into a successor library, so we should probably move to that as well.

Reply:
I switched from chobo to itlib upstream in develop yesterday, so you could use itlib small_vector

Reply:
Thanks @rob-p! So the vectors `b` and `u` tell the position of the barcode and umi in the regex and they are set only once. For eg, in the regex `([ATGC]{9,10})(CAGAGC)([ATGC]{8})([ATGC]{10})` the barcode is in group 1 and 4, and that's what `b` stores. It is shared over instances but is threadsafe because it's not modified in extractUMI and BC.

Reply:
Yes, the vector as static would be shared by every instance of the class. That is highly unsafe in a multi-threaded environment.

Reply:
Can you shed some more light on it @gmarcais? What makes it unsafe and how can we make it safe? There's also a similar question here https://stackoverflow.com/questions/39487075/local-static-variable-initialization-is-thread-safe

Reply:
@Gaura — Ok, so to be clear, when the comment says "store positions of matches for bc and umi", this is not e.g. finding the position as it occurred in each read and storing those in this class, rather, this is parsing the format string and storing the regex group index where it expects to find the bc and umi in all reads?  I _think_ if we are careful in initialization then, it may be OK (i.e. if these are read-only in all threads throughout program execution after the initialization).  Right @gmarcais?

Reply:
Ah, let me update the comment. Yes, these are read only after initialization.

Reply:
Thanks for the clarification, I misunderstood the code originally. Yes, it is safe. It behaves the same as [call_once](https://en.cppreference.com/w/cpp/thread/call_once) for the constructor.